### DISCUSSION TITLE
Run code inside anyio WorkerThread

### DISCUSSION THREAD

**Discussion by egorgam** [fastapi/fastapi#12719] [Questions] [answered]
Hello!

I trying to use FastAPI with Django ORM and since 5.1 there is builtin db ConnectionPool. I need to close db connections in anyio threads manually (because https://code.djangoproject.com/ticket/35672#comment:12), but I have no idea how can it be. Looks like only way is add my code in the end of anyio._backends._asyncio.WorkerThread.run method.

Maybe FastAPI has some way to pretty configure/patch it, or other way to run code inside auto generated threads (when app get request to sync route)?

### Operating System

Linux, macOS

### FastAPI Version

0.112.0

### Pydantic Version

2.9.2

### Python Version

3.12.7

**Comment by theoohoho** [comment]
I can't help with your question, but I have some ideas I'd like to share with you. 

If you're more familiar with Django ORM, you might consider integrating [Tortoise-ORM](https://github.com/tortoise/tortoise-orm) with FastAPI. Alternatively, if you're comfortable with Django and like the idea of FastAPI, you might try using [django-ninja](https://django-ninja.dev/).

**Comment by egorgam** [comment, marked as answer]
@theoohoho I found one way to resolve it:

1. Create class inherited by WorkerThread, and redefine run() method with original code and addition:

```python
from anyio._backends._asyncio import AsyncIOBackend, WorkerThread
from anyio._core._eventloop import claim_worker_thread, threadlocals
from django.db import close_old_connections


class CustomWorkerThread(WorkerThread):
    """
    AnyIO thread worker modified for correct close db connection.

    """

    def run(self) -> None:
        with claim_worker_thread(AsyncIOBackend, self.loop):
            while True:
                item = self.queue.get()
                if item is None:
                    return

                context, func, args, future, cancel_scope = item
                if not future.cancelled():
                    result = None
                    exception: BaseException | None = None
                    threadlocals.current_cancel_scope = cancel_scope
                    try:
                        result = context.run(func, *args)
                    except BaseException as exc:
                        exception = exc
                    finally:
                        del threadlocals.current_cancel_scope
                        close_old_connections()  # <-- gotcha

                    if not self.loop.is_closed():
                        self.loop.call_soon_threadsafe(self._report_result, future, result, exception)

                self.queue.task_done()
```

2. In main.py patch WorkerThread like here:

```python

from anyio._backends import _asyncio

_asyncio.WorkerThread = CustomWorkerThread

```

But maybe a more clean way is exists? About your recommendations - thanks, I know about Tortoise ORM and Django Ninja, but my point is use clean fastapi/pydantic with clean django orm (with async, since 5.0) https://github.com/egorgam/fango

**Reply by YuriiMotov** [reply]
You may also find these PRs helpful: #11782, #11807
